core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java

```diff
@@ -571,71 +572,100 @@ public static Connection getDBConnection(DataSource dataSource) throws SQLExcept
     }
 
     public static void closeConnection(Connection dbConnection) {
-
-        if (dbConnection != null) {
-            try {
-                dbConnection.close();
-            } catch (SQLException e) {
-                log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
-            }
-        }
+        close(dbConnection);
     }
 
-    private static void closeResultSet(ResultSet rs) {
-
-        if (rs != null) {
+    /**
+     *
+     * @param dbObject -  java.sql object types that implement both Wrapper and AutoCloseable interfaces
+     */
+    private static <AutoClosableWrapper extends AutoCloseable & Wrapper> void close(AutoClosableWrapper dbObject) {
+        if (dbObject != null) {
             try {
-                rs.close();
+                dbObject.close();
+            } catch (SQLRecoverableException ex) {
+                handleSQLRecoverableException(dbObject, ex);
             } catch (SQLException e) {
-                log.error("Database error. Could not close result set  - " + e.getMessage(), e);
+                log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
+            } catch (Exception e) {
+                log.error("An unknown error occurred during close operation" + e.getMessage(), e);
+            } finally {
+                dbObject = null;
             }
         }
-
     }
 
-    private static void closeStatement(PreparedStatement preparedStatement) {
-
-        if (preparedStatement != null) {
-            try {
-                preparedStatement.close();
-            } catch (SQLException e) {
-                log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
-            }
+    private static <AutoClosableWrapper extends AutoCloseable & Wrapper>  void handleSQLRecoverableException(AutoClosableWrapper dbObject, SQLRecoverableException recException){
```

Comment:
I think it might be better if this method name can imply that it is only handling SQLRecoverableException of databaseObject.close() operation. On the other hand I do not see a reason to retry a close operation because at the time close operation is invoked the database operation may have already completed.